[JPEGXL] Migrate JXLDecoderPtr to be a std::unique_ptr

litherum opened the pull request:
#### 02bc26dd0ddb42f260ce4aec447d4fe8278ebfb3
<pre>
[JPEGXL] Migrate JXLDecoderPtr to be a std::unique_ptr
<a href="https://bugs.webkit.org/show_bug.cgi?id=258283">https://bugs.webkit.org/show_bug.cgi?id=258283</a>
rdar://111008382

Reviewed by Cameron McCormack.

There&apos;s no need to duplicate the functionality of std::unique_ptr.

* Source/WebCore/platform/image-decoders/jpegxl/JxlDecoderPtr.h:
(JXLDecoderDeleter::operator()):
(JxlDecoderMake):
(JxlDecoderPtr::JxlDecoderPtr): Deleted.
(JxlDecoderPtr::operator=): Deleted.
(JxlDecoderPtr::~JxlDecoderPtr): Deleted.
(JxlDecoderPtr::reset): Deleted.
(JxlDecoderPtr::operator bool const): Deleted.
(JxlDecoderPtr::get const): Deleted.
(): Deleted.

Canonical link: <a href="https://commits.webkit.org/265310@main">https://commits.webkit.org/265310@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a048de743cf33c992baa543eddca63a26e5428f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10578 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10801 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11080 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12224 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10150 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10592 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13171 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10765 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13079 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10739 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11665 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8897 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12625 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8959 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9555 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16813 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10030 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9705 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12956 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10169 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8249 "3 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9324 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2527 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13587 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10026 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/platform/image-decoders/jpegxl/JxlDecoderPtr.h:
```diff
@@ -28,65 +28,24 @@
-    JxlDecoder* m_decoder { nullptr };
 };
 
+using JxlDecoderPtr = std::unique_ptr<JxlDecoder, JXLDecoderDeleter>;
```
Could we get away with using `jxl/decode_cxx.h` directly?

litherum replied:
Nope :)
1. We have a policy of not using C++ APIs from other libraries/frameworks, because of C++'s binary compatibility problems
2. AppleJPEGXL doesn't even include that header in the first place

aprotyas replied:
TIL for both, makes sense!

litherum committed:
[JPEGXL] Migrate JXLDecoderPtr to be a std::unique_ptr
https://bugs.webkit.org/show_bug.cgi?id=258283
rdar://111008382

Reviewed by Cameron McCormack.

There's no need to duplicate the functionality of std::unique_ptr.

* Source/WebCore/platform/image-decoders/jpegxl/JxlDecoderPtr.h:
(JXLDecoderDeleter::operator()):
(JxlDecoderMake):
(JxlDecoderPtr::JxlDecoderPtr): Deleted.
(JxlDecoderPtr::operator=): Deleted.
(JxlDecoderPtr::~JxlDecoderPtr): Deleted.
(JxlDecoderPtr::reset): Deleted.
(JxlDecoderPtr::operator bool const): Deleted.
(JxlDecoderPtr::get const): Deleted.
(): Deleted.

Canonical link: https://commits.webkit.org/265310@main
